File: app/src/main/java/com/zuk0/gaijinsmash/riderz/ui/shared/lifecycle/AbstractLifecycleOwner.java
package com.zuk0.gaijinsmash.riderz.ui.shared.lifecycle;

import androidx.annotation.NonNull;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;
import androidx.lifecycle.LifecycleRegistry;

public abstract class AbstractLifecycleOwner<T> implements LifecycleOwner {

    private LifecycleRegistry lifecycleRegistry;

    public AbstractLifecycleOwner(AbstractLifecycleObserver observer) {
        lifecycleRegistry = new LifecycleRegistry(this);
        getLifecycle().addObserver(observer);
    }

    public void startOwner() {
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START);
    }

    public void stopOwner() {
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP);
    }

    public void destroyOwner() {
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY);
    }

    @NonNull
    @Override
    public Lifecycle getLifecycle() {
        return lifecycleRegistry;
    }
}

File: app/src/main/java/com/zuk0/gaijinsmash/riderz/ui/shared/lifecycle/AbstractLifecycleOwner.kt
package com.zuk0.gaijinsmash.riderz.ui.shared.lifecycle

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LifecycleRegistry

abstract class AbstractLifecycleOwner<T>(observer: LifeCycleObserverContract?) : LifecycleOwner {
    private val lifecycleRegistry: LifecycleRegistry = LifecycleRegistry(this)
    fun startOwner() {
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_START)
    }

    fun stopOwner() {
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_STOP)
    }

    fun destroyOwner() {
        lifecycleRegistry.handleLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    }

    override fun getLifecycle(): Lifecycle {
        return lifecycleRegistry
    }

    init {
        lifecycle.addObserver(observer!!)
    }
}